src/main/java/com/redhat/devtools/intellij/lsp4ij/internal/CancellationSupport.java

```diff
@@ -32,7 +32,7 @@ public class CancellationSupport implements CancelChecker {
 	private boolean cancelled;
 
 	public CancellationSupport() {
-		this.futuresToCancel = new ArrayList<CompletableFuture<?>>();
+		this.futuresToCancel = new CopyOnWriteArrayList<>();

```

Comment:
I think the proper fix is to forbid the add future if it is cancelled. In other words the first line of execute m'étais should be checkCancelled()

Reply:
So a CancellationSupport is used for a single request? Because once cancelled once, then it can never be reused

Reply:
Yes it is the idea

Reply:
The main idea is to cancel previous requests for à given opération. A cancel support is used to cancel those prévious requests.

Reply:
In other words when cancel support is canceled it cannot be reused

Reply:
even if you do 
```java
	public <T> CompletableFuture<T> execute(CompletableFuture<T> future) {
		if (!cancelled) {
			this.futuresToCancel.add(future);
		}
		return future;
	}
```

callers will still join on the original future, returning null would be very inconvenient as lots of places would have to adapt

Reply:
I think when cancel support is cancelled we should cancel the future which is add whenbexecute is callled and throw a cannellableexception to stop the process

Reply:
you mean 
```java
	public <T> CompletableFuture<T> execute(CompletableFuture<T> future) {
		if (cancelled) {
			future.cancel(true);
		} else {
			this.futuresToCancel.add(future);
		}
		return future;
	}

```

where do you need to explicitly throw a cancellable exception?

Reply:
or 
```java
public <T> CompletableFuture<T> execute(CompletableFuture<T> future) {
		checkCanceled(); // throws CancellationException if cancelled
		this.futuresToCancel.add(future);
		return future;
	}
```
?

Reply:
you mean 
```java
	public <T> CompletableFuture<T> execute(CompletableFuture<T> future) {
		if (cancelled) {
			future.cancel(true);
                        checkCancelded
		} else {
			this.futuresToCancel.add(future);
		}
		return future;
	}

```

Reply:
done